hail/python/hail/methods/statgen.py

```diff
@@ -1697,6 +1697,240 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
+    mt = mt.select_entries(x = mt[field])
+    ht = mt.localize_entries(entries_array_field_name='entries')
+    ht = ht.select(xs = ht.entries.map(lambda e: e['x']))
+    ht = ht.add_index()
+    A = ht.group_by(row_group_number = hl.int32(ht.idx // block_size)) \
+        .aggregate(ndarray = hl.nd.array(hl.agg.collect(ht.xs)))
+
+    # Set Parameters
+    # block_size = 4 # might want to make this a parameter or make some heurstic for it
+    q = q_iterations
+    l = k + oversampling_param
+    n = A.take(1)[0].ndarray.shape[1]
+
+    # Generate random matrix G
+    G = np.random.normal(0, 1, (n,l))
+
+
+    # Helper Functions
+
+    def chunk_ndarray(a, group_size):
+        n_groups = ceil(a.shape[0] / group_size)
+        groups = []
+        for i in range(n_groups):
+            start = i * group_size
+            if i == n_groups - 1:
+                end = a.shape[0]
+            else:
+                end = (i + 1) * group_size
+            groups.append(a[start:end, :])
+        return groups
+
+    def concatBlocked(A):
+        blocks = A.ndarray.collect()
+        big_mat = np.concatenate(blocks, axis=0)
+        ht = ndarray_to_table([big_mat])
+        return ht
+
+    def concatToNumpy(A):
+        blocks = A.ndarray.collect()
+        big_mat = np.concatenate(blocks, axis=0)
+        return big_mat
+
+    def ndarray_to_table(chunked_arr):
+        structs = [hl.struct(row_group_number = idx, ndarray = block)
+                   for idx, block in enumerate(chunked_arr)]
+        ht = hl.Table.parallelize(structs)
+        ht = ht.key_by('row_group_number')
+        return ht
+
+    def block_product(left, right):
+        product = left @ right     
+        n_rows, n_cols = product.shape
+        return hl.struct(
+            shape=product.shape,
+            block=hl.range(hl.int(n_rows * n_cols)).map(
+                lambda absolute: product[absolute % n_rows, absolute // n_rows]))
+
+    # takes in output of block_product
+    def block_aggregate(prod):
+        shape = prod.shape
+        block = prod.block
+        return hl.nd.from_column_major(
+            hl.agg.array_sum(block),
+            hl.agg.take(shape, 1)[0])
+
+    # returns flat array
+    def to_column_major(ndarray):
+        n_rows, n_cols = ndarray.shape
+        return hl.range(hl.int(n_rows * n_cols)).map(
+            lambda absolute: ndarray[absolute % n_rows, absolute // n_rows])
+
+    def matmul_rowblocked_nonblocked(A, B):
+        temp = A.annotate_globals(mat = B)
+        temp = temp.annotate(ndarray = temp.ndarray @ temp.mat)
+        temp = temp.select(temp.ndarray)
+        temp = temp.drop(temp.mat)
+        return temp
+
+    def matmul_colblocked_rowblocked(A, B):
+        temp = A.transmute(ndarray = block_product(A.ndarray.transpose(), B[A.row_group_number].ndarray))
+        result_arr_sum = temp.aggregate(block_aggregate(temp.ndarray))
```

Comment:
Same comment as I made on the script, you no longer need the special aggregator.